### ISSUE TITLE
SmartLedsAdapter::new does not have async impl for Rmt::new_async?

### ISSUE THREAD

**Issue by brainstorm** [esp-rs/esp-hal-community#4]
Mentioned in [esp-rs matrix](https://matrix.to/#/!LdaNPfUfvefOLewEIM:matrix.org/$I_pAgUucvM7BxU2GP3FWeFJ1BePbG0vBzCV3SNH1mHQ?via=matrix.org&via=tchncs.de&via=mozilla.org) and [here too](https://matrix.to/#/!LdaNPfUfvefOLewEIM:matrix.org/$V1u-vDWg6xdCj1CTFOg1ZrHQX6i-J5BTcWSw8FVgOGA?via=matrix.org&via=tchncs.de&via=mozilla.org), flagging it here:

```rust
error[E0277]: the trait bound `esp_hal::rmt::ChannelCreator<Async, 0>: TxChannelCreator<'_, _, _>` is not satisfied
  --> src/main.rs:27:37
   |
27 |     let led = SmartLedsAdapter::new(rmt.channel0, io.pins.gpio38, rmt_buffer, &clocks);
   |               --------------------- ^^^^^^^^^^^^ the trait `TxChannelCreator<'_, _, _>` is not implemented for `esp_hal::rmt::ChannelCreator<Async, 0>`
   |               |
   |               required by a bound introduced by this call
   |
   = help: the following other types implement trait `TxChannelCreator<'d, T, P>`:
             <esp_hal::rmt::ChannelCreator<Blocking, 0> as TxChannelCreator<'d, esp_hal::rmt::Channel<Blocking, 0>, P>>
             <esp_hal::rmt::ChannelCreator<Blocking, 1> as TxChannelCreator<'d, esp_hal::rmt::Channel<Blocking, 1>, P>>
             <esp_hal::rmt::ChannelCreator<Blocking, 2> as TxChannelCreator<'d, esp_hal::rmt::Channel<Blocking, 2>, P>>
             <esp_hal::rmt::ChannelCreator<Blocking, 3> as TxChannelCreator<'d, esp_hal::rmt::Channel<Blocking, 3>, P>>
```

**Comment by robamu** [comment]
I actually implemented this before I saw the issue.

I opened a PR here:

https://github.com/esp-rs/esp-hal-community/pull/6

 I successfully used it to drive a 46 LED lightstrip asynchronously :)

**Comment by tschundler** [comment]
I wasn't able to get this working well enough for Burning Man so priority of fixing it up dropped. I'll play with your PR when I get a chance. I have a variety of LEDs with different timing tolerances.

Skimming it... So you didn't work around the bug that async RMT right now can't refill the buffer and only supports one-shot tx, rather you're sending a bunch of single shorts. As long as the next single shot is less than the latch time, that would work. That's simpler than my approach trying to fix RMT to allow arbitrary input lengths in async mode. (And then doing to conversion to pulse-codes on the fly)

Have you checked how long it is between chunks? I wonder if it could do the pulse code conversion while RMT is sending in the SmartLEDs code. My goals are parallel output and lower memory usage to support >2k total LEDs.

**Comment by robamu** [comment]
I actually measure the time compared to the blocking API. Writing all 46 LEDs with async API takes around 3-4 ms, blocking API takes ~2 ms (but is blocking, which incurred issues with the IR receiver handling) . Considering than my LED fine ticker timer is 20 ms, this was not a problem for me. I actually saw that the blocking API is doing some magic of continously filling the other buffer, and doing that with the async API would probably require changes or a custom driver, because the async API does not listen to the threshold interrupts.

> 2k LEDs sounds massive.

**Comment by robamu** [comment]
Each write takes around 52 microseconds.

**Comment by tschundler** [comment]
That's close to what I got testing your code. It work with only 2 of 6 types of LEDs that I have handy. For others, the extra 20-some microsecond delay between pixels is treated as the latch time, and you only get one pixel working.

So I think it is reasonable with comments and we can work on improving it.

**Comment by robamu** [comment]
Hmm, does that mean that these pixels would require a new transfer to be started immediately? The current async API does not allow this from what I can see, custom (interrupt) code would probably required to do this :/

**Comment by tschundler** [comment]
> Hmm, does that mean that these pixels would require a new transfer to be started immediately? The current async API does not allow this from what I can see, custom (interrupt) code would probably required to do this :/

The new transfer needs to be within 6us judging by https://wp.josh.com/2014/05/13/ws2812-neopixels-are-not-so-finicky-once-you-get-to-know-them/amp/
>  If this gap is longer than ~6,000ns, it will become a reset pulse (TLL) to the next pixel in the string.

despite the datasheet saying 50us for latch time. Though it seems some clones do have a longer latch time.

**Comment by robamu** [comment]
Okay, that's unfortunate. Just so I understand correctly: The amount of time required for the RMT to finish, wake the task again, context switch back to start the next transfer is longer than those 6 us? I don't see another reliable way to deal with this other than interrupts to blocking/polling just like the blocking API does..

**Comment by tschundler** [comment]
> The amount of time required for the RMT to finish, wake the task again, context switch back to start the next transfer is longer than those 6 us?

Yes, it looks like it is taking 20uS (~30us to send 24 bits + ~50uS total per pixel)

I found that when I faked async by doing a loop of polling I had no problem, and I forgot how, but I estimated the async context switch to take about 8us, which seemed long, but I was also suspicious that I'm doing something incorrectly.

---

I've realized that testing my variant, I should probably flip pins for when ISR is running and when the task is awake, then I can see the timing relationship better.

(And in my case, since my real goal is parallel output, maybe I2S/LCD out is what I should really be looking into)
